### PR TITLE
Add turbo:fetch-request-error event on frame and form network errors

### PR DESCRIPTION
**Pull request by srt32**
closes https://github.com/hotwired/turbo/issues/462 and possible https://github.com/hotwired/turbo/issues/460

This PR adds a new event called `turbo:fetch-error` that dispatches
when a form or frame fetch request errors.

This event will let developers respond appropriately when a fetch request
fails due network errors (such as on flaky wifi). A developer could reload the page or show an error message of some kind.

At the moment, if a frame navigation fetch request fails due to a network
error an error is thrown so https://github.com/hotwired/turbo/blob/2d5cdda4c030658da21965cb20d2885ca7c3e127/src/http/fetch_request.ts#L107 never runs
and therefore the `turbo:before-fetch-response` event is not dispatched.

Here is a demo using http://127.0.0.1:9000/src/tests/fixtures/tabs.html as a fixture:

https://user-images.githubusercontent.com/2181356/179891552-86e46b86-6d2e-4e3e-a1da-1066dea44550.mp4

I'm very open to feedback or other ideas on a good direction to go here.

### CODE REVIEW

File: src/core/drive/form_submission.ts [hunk 1]
```diff
@@ -199,6 +200,11 @@ export class FormSubmission {
 
   requestErrored(request: FetchRequest, error: Error) {
     this.result = { success: false, error }
+    dispatch<TurboFetchErrorEvent>("turbo:fetch-error", {
+      target: this.formElement,
+      detail: { request, error },
+      cancelable: true,
```
Comment (keithamus) [inline src/core/drive/form_submission.ts:206]: This probably shouldn't be cancelable. 

```suggestion
```

File: src/core/frames/frame_controller.ts [hunk 2]
```diff
@@ -241,6 +241,11 @@ export class FrameController
 
   requestErrored(request: FetchRequest, error: Error) {
     console.error(error)
+    dispatch<TurboFetchErrorEvent>("turbo:fetch-error", {
+      target: this.element,
+      detail: { request, error },
+      cancelable: true,
```
Comment (keithamus) [inline src/core/frames/frame_controller.ts:247]: This probably shouldn't be cancelable. 

```suggestion
```

**Comment by seanpdoyle** [comment]
Thank you for proposing this! Is it possible to add test coverage? Does playwright have the ability to force the Chrome or Firefox browser running behind the scenes into Offline mode?

**Comment by srt32** [comment]
> Thank you for proposing this! Is it possible to add test coverage? Does playwright have the ability to force the Chrome or Firefox browser running behind the scenes into Offline mode?

I've got something ugly in https://github.com/hotwired/turbo/pull/640/commits/ecdd1c9a054914f19e69f013c94941f46a4c423a and curious if you know of a good way to determine which browser we need to pull in.

File: src/tests/functional/frame_navigation_tests.ts [hunk 3]
```diff
@@ -24,6 +25,25 @@ test("test frame navigation with exterior link", async ({ page }) => {
...
+  if (testInfo.project.name === "chrome") {
+    browser = await chromium.launch()
+  } else if (testInfo.project.name === "firefox") {
+    browser = await firefox.launch()
+  } else {
+    throw new Error("unsupported browser")
+  }
+  const context = await browser.newContext()
+  page = await context.newPage()
+  await context.setOffline(true)
```
Comment (seanpdoyle) [inline src/tests/functional/frame_navigation_tests.ts:40]: Could we get at this `Context` instance through [Page.context()](https://playwright.dev/docs/api/class-page#page-context)?

File: src/tests/functional/frame_navigation_tests.ts [hunk 4]
```diff
@@ -24,6 +24,13 @@ test("test frame navigation with exterior link", async ({ page }) => {
   await nextEventOnTarget(page, "frame", "turbo:frame-load")
 })
 
+test("test frame navigation emits fetch-error event when offline", async ({ page }) => {
+  await page.goto("/src/tests/fixtures/tabs.html")
```
Comment (seanpdoyle) [inline src/tests/functional/frame_navigation_tests.ts:28]: While I appreciate that you're recreating the failure you experienced with the page that you experienced it on, is there a link in the `/src/tests/fixtures/frame_navigation.html` page that we could exercise for the same outcome? That way we could cut down on the number of `page.goto` calls.

**Review by seanpdoyle — APPROVED** [review]

File: src/core/frames/frame_controller.ts [hunk 5]
```diff
@@ -247,6 +247,10 @@ export class FrameController
 
   requestErrored(request: FetchRequest, error: Error) {
     console.error(error)
+    dispatch<TurboFetchErrorEvent>("turbo:fetch-error", {
```
Comment (seanpdoyle) [inline src/core/frames/frame_controller.ts:250]: How do you feel about naming the event `turbo:fetch-request-error`? Other [fetch-related events](https://turbo.hotwired.dev/reference/events) are named to describe which side of they pipe they're dispatched on: `turbo:before-fetch-request` and `turbo:before-fetch-response`.

**Review by seanpdoyle — APPROVED** [review]

File: src/core/frames/frame_controller.ts (hunk 5 above)
Comment (seanpdoyle) [reply src/core/frames/frame_controller.ts:250]: Also, could you add the type to the [exports from `src/core/index.ts`](https://github.com/hotwired/turbo/blob/60c646383b3c0355242d2e3c69f80742efb266be/src/core/index.ts#L17-L31)?

Comment (srt32) [reply src/core/frames/frame_controller.ts]: Done in db678ec176d74eff9c82876e05b979b1fddc16c2! I'm still fighting tests a bit.

File: src/tests/functional/frame_navigation_tests.ts [hunk 6]
```diff
@@ -23,6 +23,13 @@ test("test frame navigation with exterior link", async ({ page }) => {
   await nextEventOnTarget(page, "frame", "turbo:frame-load")
 })
 
+test("test frame navigation emits fetch-request-error event when offline", async ({ page }) => {
+  await page.goto("/src/tests/fixtures/tabs.html")
+  await page.context().setOffline(true)
+  await page.click("#tab-2")
+  await nextEventNamed(page, "turbo:fetch-request-error")
```
Comment (seanpdoyle) [inline src/tests/functional/frame_navigation_tests.ts:30]: @srt32 to use this helper method, you need to add an entry for the event name to the src/tests/test.js file.

**Comment by srt32** [comment]
@seanpdoyle thanks for your help here! This change should be green and ready to go. I bumped into some recursion issues with the test helper so I opted to not serialize the `detail` for that specific event. I'm not sure exactly why this event is behaving differently than the others but the build is 🟢  now ;)

**Review by seanpdoyle — APPROVED** [review]

**Comment by srt32** [comment]
Thanks for the thumb! What do we need to do to get this PR merged?

**Pull request merged**